pretty print: use single-char square root symbol in simple cases

cbm755 opened the pull request:
For single-char Symbols and for non-negative integers we use
the unicode square root symbol.  As this doesn't include a top
overline indicated what the root encompasses, we use this only
for simple unambiguous cases like:

```
>>> pprint(exp(3*sqrt(5*x*pi)))
 3⋅√5⋅√π⋅√x
ℯ
```

Update tests for this.

Perhaps slightly more controversial:

```
>>> pprint(sqrt(123))
√123
>>> pprint(sqrt(Symbol('C1')))
  ____
╲╱ C₁
```

Note---as in this last case---we do _not_ simply use "√" for all
Symbols because "√C₁" looks ambiguous.  Add test for this.

An alternative would be to combine the "√" symbol with some
overlining, although at some cost in vertical space.  And the
overline does not connect to the "√" symbol.

cbm755 committed:
pretty print: use single-char square root symbol in simple cases

For single-char Symbols and for non-negative integers we use
the unicode square root symbol.  As this doesn't include a top
overline indicated what the root encompasses, we use this only
for simple unambiguous cases like:
````
>>> pprint(exp(3*sqrt(5*x*pi)))
 3⋅√5⋅√π⋅√x
ℯ
````
Update tests for this.

Perhaps slightly more controversial:
````
>>> pprint(sqrt(123))
√123
>>> pprint(sqrt(Symbol('C1')))
  ____
╲╱ C₁
````
Note---as in this last case---we do *not* simply use "√" for all
Symbols because "√C₁" looks ambiguous.  Add test for this.

An alternative would be to combine the "√" symbol with some
overlining, although at some cost in vertical space.  And the
overline does not connect to the "√" symbol.

cbm755 committed:
pretty print: doc fixes for single-char square root symbol

moorepants commented:
```
>>> pprint(sqrt(123))
√123
```

I'm not sure multi-digit numbers should use this radical symbol. It can be slightly misleading. I had to look twice to think about what it was. The overline is unambiguous.

cbm755 commented:
Sure, that makes sense.  Would you like "√2" but keep status quo for `sqrt(123)`?

moorepants commented:
Yes, the non-overline for single digits seems nice and it is not ambiguous.

toolforger commented:
Even ```√5⋅√π``` is ambiguous - it could be misinterpreted as ```sqrt(5*(sqrt(pi))```.
My preference would be to be strictly unambigous, which means an overline.
Or restrict the square root symbol to the last term before a closing parentheses or end-of-expression, i.e. allow ```√5``` and ```...√5)``` but do not do ```... √5⋅...```, and probably not even ```... √5 + ...```.

cbm755 commented:
For the record, here is another (currently unimplemented) alternative, but at least in the font I have it doesn't really look that great:
````
 ___
√123
````

toolforger commented:
Or maybe do ```(√5)```. That's always unambiguous (but it's also ugly).

moorepants commented on doc/src/tutorial/intro.rst:
```diff
@@ -163,21 +161,17 @@ Find the eigenvalues of `\left[\begin{smallmatrix}1 & 2\\2 &
- ⎨─ + ──────: 1, - ────── + ─: 1⎬
- ⎪2     2            2      2   ⎪
- ⎩                              ⎭
+ ⎧3   √17       √17   3   ⎫
```
This one doesn't look so bad though, compared to the three digit one we discussed in the pr comments. Hmm. I guess if SymPy always puts a multiplication dot between values then maybe the non-overline would work.

moorepants commented:
> Even ```√5⋅√π``` is ambiguous - it could be misinterpreted as ```sqrt(5*(sqrt(pi))```.

I think that misinterpretation is a stretch.

certik commented:
@cbm755 this is awesome, thanks! 

I would add it into our options --- I think there is quite a few options that the user can set for the pretty printer (ascii vs unicode, etc.), so one of them should be `unicode_square_roots`. I think it can be on by default.

cbm755 commented:
@certik thanks.  I'm not a huge fan of adding knobs but if you want I can look into it.

certik commented:
The advantage of being able to turn this off is if it causes problems, then people can turn it off.

cbm755 committed:
pretty: square root: add setting for unicode square root symbol

Also docs and tests for setting.

cbm755 commented:
Ok @certik  I added a setting.  Let me know if its ok, or needs different name, etc.

certik commented:
Thanks @cbm755, I think it works great!
```
In [19]: print pretty(sqrt(2*5*6*7*878787*54654654)*x, unicode_square_roots=False)
      _______________  
126⋅╲╱ 1270629614410 ⋅x

In [20]: print pretty(sqrt(2*5*6*7*878787*54654654)*x)
126⋅√1270629614410⋅x
```
I think this way it is safe --- if it causes problems, people can turn it off immediately for the given expression. I think the way it looks is better, so it should be on by default.

However, you have a test failure there:
```
sympy/printing/pretty/tests/test_pretty.py[0] Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/travis/virtualenv/python3.2.5/lib/python3.2/site-packages/sympy-0.7.6_git-py3.2.egg/sympy/utilities/runtests.py", line 512, in _test
    slow=slow, enhance_asserts=enhance_asserts))
  File "/home/travis/virtualenv/python3.2.5/lib/python3.2/site-packages/sympy-0.7.6_git-py3.2.egg/sympy/utilities/runtests.py", line 964, in test
    self.test_file(f, sort, timeout, slow, enhance_asserts)
  File "/home/travis/virtualenv/python3.2.5/lib/python3.2/site-packages/sympy-0.7.6_git-py3.2.egg/sympy/utilities/runtests.py", line 1028, in test_file
    code = compile(source, filename, "exec")
  File "/home/travis/virtualenv/python3.2.5/lib/python3.2/site-packages/sympy-0.7.6_git-py3.2.egg/sympy/printing/pretty/tests/test_pretty.py", line 1719
    """
      ^
SyntaxError: invalid syntax
Traceback (most recent call last):
  File "<stdin>", line 3, in <module>
Exception: Tests failed
```
This happens in Python 3.2 only and the reason is that it doesn't support the `u` in the string --- you need to use `u()`, see the other tests.

cbm755 committed:
pretty print: rename sqrt symbol option, clean up tests

Rename the option to "use_unicode_sqrt_char".
Also, I had a new test function right in the middle of another
test function.  Move the test around.  Split off the "C1" test
into its own test fcn.

cbm755 commented:
oops, sorry about that.  Fixed that up.  I renamed the option to `use_unicode_sqrt_char`.  Cleaned up tests a bit.

@certik @moorepants if you're both +1, then I'll merge this.

cbm755 commented:
After merge, I will paste this into release notes:

    The unicode pretty printer now uses a compact single-character for simple
    expressions like `sqrt(x)` or `sqrt(17)`.  You can disable this new behaviour
    with `pprint(sqrt(2), use_unicode_sqrt_char=False)`.

certik commented:
+1 to merge if tests pass.

Sent from my mobile phone.
On Apr 5, 2015 4:21 AM, "Colin Macdonald" <notifications@github.com> wrote:

> After merge, I will paste this into release notes:
>
> The unicode pretty printer now uses a compact single-character for simple
> expressions like `sqrt(x)` or `sqrt(17)`.  You can disable this new behaviour
> with `pprint(sqrt(2), use_unicode_sqrt_char=False)`.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sympy/sympy/pull/9234#issuecomment-89748956>.
>

moorepants commented:
+1

asmeurer commented:
I just noticed this change. I'm not a huge fan. I get the desire to make pretty printed expressions smaller, but this makes things harder to read IMO.

asmeurer commented:
Did you look at using a Unicode combining overline at all? It would look like this

```
√2̅
```

The bad news is that, at least in my browser, √2̅ renders incorrectly in the code section of this comment(but fine in the non-code section). It looks like this in my terminal.

<img width="248" alt="screenshot 2015-07-07 15 14 29" src="https://cloud.githubusercontent.com/assets/71486/8556169/e78fc8a4-24ba-11e5-89d2-0d517416ef11.png">

In general, though, this seems squished, but that could just be me being used to the old way.  But I think in general SymPy's pretty printer is generous with whitespace (notice how there is extra space to the left and right of `x + 1` in the example below. Using `x²` for `x**2` always felt the same. 

It also seems a little odd to have this inconsistency:

```
In [7]: sqrt(x + 1)
Out[7]:
  _______
╲╱ x + 1

In [8]: sqrt(x)
Out[8]: √x
```

but then again, the SymPy pretty printer is already pretty dynamic (e.g., using parentheses only when necessary for summations and integrals).

certik commented:
Fortunately we can turn this on and off, so I am ok wither way as default.

cbm755 commented:
@asmeurer first of all, sorry.

I didn't play with that combining overline.  On my terminal (Gnome 3 on Fedora 22) they don't seem to combine property, but I haven't looked carefully (just pasted in your example).  I had some other bug that I recently fixed about combining characters so I don't trust them much :(

I like this as is, but of course I also don't mind if the default was the other way.

cbm755 commented:
Slightly off-topic but I personally don't like the regular bit sqrts because they combine underscores with box drawing---they have the wrong weight in my font which looks odd when zoomed in

![screenshot from 2015-07-07 21-29-19](https://cloud.githubusercontent.com/assets/818622/8556504/4fb629da-24ef-11e5-9172-2f78d70a39bb.png)

(looking forward to someday owning a laptop for which that is my actual font size!)
